### PR TITLE
Behold the arse merkle tree

### PR DESCRIPTION
**Pull request by batconjurer**
Merging in the arse merkle tree

### CODE REVIEW

File: shared/src/ledger/storage/merkle_tree.rs [hunk 1]
```diff
@@ -1,44 +1,63 @@
...
+use arse_merkle_tree::{
+    Hash as SmtHash, Key as TreeKey, SparseMerkleTree as ArseMerkleTree, H256,
+};
 use borsh::{BorshDeserialize, BorshSerialize};
 use ics23::commitment_proof::Proof as Ics23Proof;
 use ics23::{
     CommitmentProof, ExistenceProof, HashOp, LeafOp, LengthOp,
     NonExistenceProof, ProofSpec,
 };
+use itertools::Either;
```
Comment (sug0) [inline shared/src/ledger/storage/merkle_tree.rs:19]: maybe switch to [this `either`](https://docs.rs/either/latest/either/)? it has some nice methods, such as [`map_left`](https://docs.rs/either/latest/either/enum.Either.html#method.map_left)

Comment (sug0) [reply shared/src/ledger/storage/merkle_tree.rs:19]: it is a relatively small dep, with optional `serde` encoding which can be disabled so build times are smaller

File: shared/src/ledger/storage/merkle_tree.rs [hunk 2]
```diff
@@ -156,81 +253,98 @@ impl<H: StorageHasher + Default> MerkleTree<H> {
...
+        let value = match self.tree(&store_type) {
+            Either::Left(smt) => {
+                smt.get(&H::hash(sub_key.to_string()).into())?.is_zero()
+            }
+            Either::Right(amt) => {
+                let key =
+                    StringKey::try_from_bytes(sub_key.to_string().as_bytes())?;
+                amt.get(&key)?.is_zero()
+            }
+        };
```
Comment (sug0) [inline shared/src/ledger/storage/merkle_tree.rs:308]: `Either::map_left` and `Either::map_right`, then `Either::into_inner`, to get a `Result<bool>`

File: shared/src/ledger/storage/merkle_tree.rs [hunk 3]
```diff
@@ -241,46 +355,76 @@ impl<H: StorageHasher + Default> MerkleTree<H> {
...
+                match cp.proof.expect("The proof should exist") {
+                    Ics23Proof::Exist(ep) => CommitmentProof {
+                        proof: Some(Ics23Proof::Exist(ExistenceProof {
+                            leaf: Some(self.ibc_leaf_spec()),
+                            ..ep
+                        })),
+                    },
+                    _ => unreachable!(),
+                }
+            }
```
Comment (sug0) [inline shared/src/ledger/storage/merkle_tree.rs:391]: same thing here, with the maps?

**Review by sug0 — CHANGES_REQUESTED** [review]

**Review by sug0 — APPROVED** [review]

**Review by james-chf — APPROVED** [review]

**Comment by batconjurer** [comment]
pls update wasm

**Pull request merged**